File: src/MarkAgent.Host/Apis/StatisticsApi.cs
using MarkAgent.Host.Domain.Services;
using Microsoft.AspNetCore.Mvc;

namespace MarkAgent.Host.Apis;

/// <summary>
/// 统计数据API
/// </summary>
[ApiController]
[Route("api/statistics")]
public class StatisticsApi : ControllerBase
{
    private readonly IToolStatisticsService _statisticsService;

    public StatisticsApi(IToolStatisticsService statisticsService)
    {
        _statisticsService = statisticsService;
    }

    /// <summary>
    /// 获取概览统计数据（用于仪表板卡片）
    /// </summary>
    [HttpGet("overview")]
    public async Task<ActionResult<OverviewStatistics>> GetOverviewStatistics()
    {
        try
        {
            var overview = await _statisticsService.GetOverviewStatisticsAsync();
            return Ok(overview);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取工具使用趋势数据（用于趋势图表）
    /// </summary>
    [HttpGet("trend")]
    public async Task<ActionResult<List<TrendDataPoint>>> GetUsageTrend([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var trendData = await _statisticsService.GetUsageTrendDataAsync(days);
            return Ok(trendData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取工具使用分布数据（用于饼图/环形图）
    /// </summary>
    [HttpGet("distribution")]
    public async Task<ActionResult<List<ToolUsageDistribution>>> GetUsageDistribution([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var distribution = await _statisticsService.GetToolUsageDistributionAsync(days);
            return Ok(distribution);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取工具成功率统计数据（用于条形图）
    /// </summary>
    [HttpGet("success-rate")]
    public async Task<ActionResult<List<ToolSuccessRateStats>>> GetSuccessRateStats([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var successRateStats = await _statisticsService.GetToolSuccessRateStatsAsync(days);
            return Ok(successRateStats);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取工具性能统计数据（用于性能图表）
    /// </summary>
    [HttpGet("performance")]
    public async Task<ActionResult<List<ToolPerformanceStats>>> GetPerformanceStats([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var performanceStats = await _statisticsService.GetToolPerformanceStatsAsync(days);
            return Ok(performanceStats);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取最近活动数据（用于活动时间线）
    /// </summary>
    [HttpGet("recent-activities")]
    public async Task<ActionResult<List<RecentActivityItem>>> GetRecentActivities([FromQuery] int count = 20)
    {
        try
        {
            if (count < 1 || count > 100)
                return BadRequest(new { error = "Count must be between 1 and 100" });

            var activities = await _statisticsService.GetRecentActivitiesAsync(count);
            return Ok(activities);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取使用热力图数据（按小时统计使用情况）
    /// </summary>
    [HttpGet("heatmap")]
    public async Task<ActionResult<Dictionary<int, Dictionary<int, int>>>> GetUsageHeatmap([FromQuery] int days = 7)
    {
        try
        {
            if (days < 1 || days > 30)
                return BadRequest(new { error = "Days must be between 1 and 30" });

            var heatmapData = await _statisticsService.GetUsageHeatmapDataAsync(days);
            return Ok(heatmapData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取工具使用排行榜
    /// </summary>
    [HttpGet("ranking")]
    public async Task<ActionResult<List<ToolRankingItem>>> GetToolRanking([FromQuery] int topCount = 10, [FromQuery] int days = 30)
    {
        try
        {
            if (topCount < 1 || topCount > 50)
                return BadRequest(new { error = "TopCount must be between 1 and 50" });
            
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var ranking = await _statisticsService.GetToolRankingAsync(topCount, days);
            return Ok(ranking);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取所有统计数据（一次性获取，减少前端请求次数）
    /// </summary>
    [HttpGet("dashboard")]
    public async Task<ActionResult<object>> GetDashboardData([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var overview = await _statisticsService.GetOverviewStatisticsAsync();
            var trendData = await _statisticsService.GetUsageTrendDataAsync(days);
            var distribution = await _statisticsService.GetToolUsageDistributionAsync(days);
            var recentActivities = await _statisticsService.GetRecentActivitiesAsync(10);

            var dashboardData = new
            {
                overview,
                trendData,
                distribution,
                recentActivities,
                lastUpdated = DateTime.UtcNow
            };

            return Ok(dashboardData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // =================== 客户端连接统计API接口 ===================

    /// <summary>
    /// 获取客户端连接概览统计
    /// </summary>
    [HttpGet("clients/overview")]
    public async Task<ActionResult<ClientOverviewStatistics>> GetClientOverviewStatistics()
    {
        try
        {
            var overview = await _statisticsService.GetClientOverviewStatisticsAsync();
            return Ok(overview);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取客户端连接趋势数据
    /// </summary>
    [HttpGet("clients/trend")]
    public async Task<ActionResult<List<ClientTrendDataPoint>>> GetClientConnectionTrend([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var trendData = await _statisticsService.GetClientConnectionTrendAsync(days);
            return Ok(trendData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取客户端分布数据
    /// </summary>
    [HttpGet("clients/distribution")]
    public async Task<ActionResult<List<ClientDistribution>>> GetClientDistribution([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var distribution = await _statisticsService.GetClientDistributionAsync(days);
            return Ok(distribution);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取客户端版本分布数据
    /// </summary>
    [HttpGet("clients/version-distribution")]
    public async Task<ActionResult<List<ClientVersionDistribution>>> GetClientVersionDistribution([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var versionDistribution = await _statisticsService.GetClientVersionDistributionAsync(days);
            return Ok(versionDistribution);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取最近的客户端连接活动
    /// </summary>
    [HttpGet("clients/recent-activities")]
    public async Task<ActionResult<List<RecentClientActivity>>> GetRecentClientActivities([FromQuery] int count = 20)
    {
        try
        {
            if (count < 1 || count > 100)
                return BadRequest(new { error = "Count must be between 1 and 100" });

            var activities = await _statisticsService.GetRecentClientActivitiesAsync(count);
            return Ok(activities);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取客户端连接排行榜
    /// </summary>
    [HttpGet("clients/ranking")]
    public async Task<ActionResult<List<ClientRankingItem>>> GetClientRanking([FromQuery] int topCount = 10, [FromQuery] int days = 30)
    {
        try
        {
            if (topCount < 1 || topCount > 50)
                return BadRequest(new { error = "TopCount must be between 1 and 50" });
            
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            var ranking = await _statisticsService.GetClientRankingAsync(topCount, days);
            return Ok(ranking);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取客户端连接热力图数据
    /// </summary>
    [HttpGet("clients/heatmap")]
    public async Task<ActionResult<Dictionary<int, Dictionary<int, Dictionary<string, int>>>>> GetClientConnectionHeatmap([FromQuery] int days = 7)
    {
        try
        {
            if (days < 1 || days > 30)
                return BadRequest(new { error = "Days must be between 1 and 30" });

            var heatmapData = await _statisticsService.GetClientConnectionHeatmapAsync(days);
            return Ok(heatmapData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    /// <summary>
    /// 获取包含客户端统计的完整仪表板数据
    /// </summary>
    [HttpGet("dashboard-with-clients")]
    public async Task<ActionResult<object>> GetDashboardWithClientsData([FromQuery] int days = 30)
    {
        try
        {
            if (days < 1 || days > 365)
                return BadRequest(new { error = "Days must be between 1 and 365" });

            // 工具统计
            var toolOverview = await _statisticsService.GetOverviewStatisticsAsync();
            var toolTrendData = await _statisticsService.GetUsageTrendDataAsync(days);
            var toolDistribution = await _statisticsService.GetToolUsageDistributionAsync(days);
            var toolRecentActivities = await _statisticsService.GetRecentActivitiesAsync(10);

            // 客户端统计
            var clientOverview = await _statisticsService.GetClientOverviewStatisticsAsync();
            var clientTrendData = await _statisticsService.GetClientConnectionTrendAsync(days);
            var clientDistribution = await _statisticsService.GetClientDistributionAsync(days);
            var clientVersionDistribution = await _statisticsService.GetClientVersionDistributionAsync(days);
            var clientRecentActivities = await _statisticsService.GetRecentClientActivitiesAsync(10);
            var clientRanking = await _statisticsService.GetClientRankingAsync(10, days);

            var dashboardData = new
            {
                tools = new
                {
                    overview = toolOverview,
                    trendData = toolTrendData,
                    distribution = toolDistribution,
                    recentActivities = toolRecentActivities
                },
                clients = new
                {
                    overview = clientOverview,
                    trendData = clientTrendData,
                    distribution = clientDistribution,
                    versionDistribution = clientVersionDistribution,
                    recentActivities = clientRecentActivities,
                    ranking = clientRanking
                },
                lastUpdated = DateTime.UtcNow
            };

            return Ok(dashboardData);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }
}

File: web/src/lib/statistics-api.ts
// 统计数据类型定义
export interface OverviewStatistics {
  totalUsageCount: number
  activeToolsCount: number
  averageSuccessRate: number
  averageExecutionTime: number
  todayUsageCount: number
  todaySuccessRate: number
  uniqueSessions: number
  mostUsedTool: string
}

export interface TrendDataPoint {
  date: string
  totalUsage: number
  successCount: number
  failureCount: number
  successRate: number
  averageExecutionTime: number
}

export interface ToolUsageDistribution {
  toolName: string
  usageCount: number
  percentage: number
  color: string
}

export interface ToolSuccessRateStats {
  toolName: string
  totalUsageCount: number
  successCount: number
  failureCount: number
  successRate: number
}

export interface ToolPerformanceStats {
  toolName: string
  averageExecutionTime: number
  minExecutionTime: number
  maxExecutionTime: number
  medianExecutionTime: number
  totalCalls: number
}

export interface RecentActivityItem {
  toolName: string
  timestamp: string
  isSuccess: boolean
  executionTime: number
  sessionId?: string
  errorMessage?: string
}

export interface ToolRankingItem {
  rank: number
  toolName: string
  usageCount: number
  successRate: number
  averageExecutionTime: number
  trendChange: number
}

export interface DashboardData {
  overview: OverviewStatistics
  trendData: TrendDataPoint[]
  distribution: ToolUsageDistribution[]
  successRateStats: ToolSuccessRateStats[]
  performanceStats: ToolPerformanceStats[]
  recentActivities: RecentActivityItem[]
  ranking: ToolRankingItem[]
  lastUpdated: string
}

// 客户端统计数据类型定义
export interface ClientOverviewStatistics {
  totalConnections: number
  activeClientsCount: number
  averageConnectionSuccessRate: number
  averageConnectionDuration: number
  todayConnectionsCount: number
  todaySuccessRate: number
  uniqueSessionsToday: number
  mostActiveClient: string
}

export interface ClientTrendDataPoint {
  date: string
  clientConnections: Record<string, number>
  totalConnections: number
  successfulConnections: number
  failedConnections: number
  successRate: number
  averageConnectionDuration: number
}

export interface ClientDistribution {
  clientName: string
  connectionCount: number
  percentage: number
  color: string
}

export interface ClientVersionDistribution {
  clientName: string
  version: string
  connectionCount: number
  percentage: number
  color: string
}

export interface RecentClientActivity {
  clientName: string
  clientVersion?: string
  connectionTime: string
  status: 'Connected' | 'Disconnected' | 'Failed' | 'Timeout'
  connectionDuration?: number
  sessionId: string
  toolUsageCount: number
  errorMessage?: string
}

export interface ClientRankingItem {
  rank: number
  clientName: string
  clientVersion?: string
  connectionCount: number
  successRate: number
  averageConnectionDuration: number
  totalToolUsages: number
  trendChange: number
}

export interface CombinedDashboardData {
  tools: {
    overview: OverviewStatistics
    trendData: TrendDataPoint[]
    distribution: ToolUsageDistribution[]
    successRateStats: ToolSuccessRateStats[]
    recentActivities: RecentActivityItem[]
    ranking: ToolRankingItem[]
  }
  clients: {
    overview: ClientOverviewStatistics
    trendData: ClientTrendDataPoint[]
    distribution: ClientDistribution[]
    versionDistribution: ClientVersionDistribution[]
    recentActivities: RecentClientActivity[]
    ranking: ClientRankingItem[]
  }
  lastUpdated: string
}

// Channel统计数据类型定义
export interface ChannelStatistics {
  pendingEventCount: number
  totalProcessedEvents: number
  totalFailedEvents: number
  lastProcessedTime: string
  isHealthy: boolean
}

export interface ChannelHealthCheck {
  status: 'healthy' | 'unhealthy'
  timestamp: string
  details: {
    pendingEvents: number
    totalProcessed: number
    totalFailed: number
    lastProcessed: string
    successRate: number
  }
}


// API客户端类
export class StatisticsApi {
  private baseUrl: string

  constructor(baseUrl: string = "") {
    this.baseUrl = baseUrl
  }

  private async request<T>(endpoint: string): Promise<T> {
    try {
      const response = await fetch(`${this.baseUrl}/api/statistics${endpoint}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      })

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }

      return await response.json()
    } catch (error) {
      console.error(`API request failed for ${endpoint}:`, error)
      throw error
    }
  }

  // 获取概览统计数据
  async getOverviewStatistics(): Promise<OverviewStatistics> {
    return this.request<OverviewStatistics>('/overview')
  }

  // 获取使用趋势数据
  async getUsageTrend(days: number = 30): Promise<TrendDataPoint[]> {
    return this.request<TrendDataPoint[]>(`/trend?days=${days}`)
  }

  // 获取工具使用分布
  async getUsageDistribution(days: number = 30): Promise<ToolUsageDistribution[]> {
    return this.request<ToolUsageDistribution[]>(`/distribution?days=${days}`)
  }

  // 获取成功率统计
  async getSuccessRateStats(days: number = 30): Promise<ToolSuccessRateStats[]> {
    return this.request<ToolSuccessRateStats[]>(`/success-rate?days=${days}`)
  }

  // 获取性能统计
  async getPerformanceStats(days: number = 30): Promise<ToolPerformanceStats[]> {
    return this.request<ToolPerformanceStats[]>(`/performance?days=${days}`)
  }

  // 获取最近活动
  async getRecentActivities(count: number = 20): Promise<RecentActivityItem[]> {
    return this.request<RecentActivityItem[]>(`/recent-activities?count=${count}`)
  }

  // 获取使用热力图数据
  async getUsageHeatmap(days: number = 7): Promise<Record<number, Record<number, number>>> {
    return this.request<Record<number, Record<number, number>>>(`/heatmap?days=${days}`)
  }

  // 获取工具排行榜
  async getToolRanking(topCount: number = 10, days: number = 30): Promise<ToolRankingItem[]> {
    return this.request<ToolRankingItem[]>(`/ranking?topCount=${topCount}&days=${days}`)
  }

  // 获取仪表板所有数据（推荐使用，减少请求次数）
  async getDashboardData(days: number = 30): Promise<DashboardData> {
    return this.request<DashboardData>(`/dashboard?days=${days}`)
  }

  // =================== 客户端统计相关API方法 ===================

  // 获取客户端连接概览统计
  async getClientOverviewStatistics(): Promise<ClientOverviewStatistics> {
    return this.request<ClientOverviewStatistics>('/clients/overview')
  }

  // 获取客户端连接趋势
  async getClientConnectionTrend(days: number = 30): Promise<ClientTrendDataPoint[]> {
    return this.request<ClientTrendDataPoint[]>(`/clients/trend?days=${days}`)
  }

  // 获取客户端连接分布
  async getClientDistribution(days: number = 30): Promise<ClientDistribution[]> {
    return this.request<ClientDistribution[]>(`/clients/distribution?days=${days}`)
  }

  // 获取客户端版本分布
  async getClientVersionDistribution(days: number = 30): Promise<ClientVersionDistribution[]> {
    return this.request<ClientVersionDistribution[]>(`/clients/version-distribution?days=${days}`)
  }

  // 获取最近客户端活动
  async getRecentClientActivities(count: number = 20): Promise<RecentClientActivity[]> {
    return this.request<RecentClientActivity[]>(`/clients/recent-activities?count=${count}`)
  }

  // 获取客户端排行榜
  async getClientRanking(topCount: number = 10, days: number = 30): Promise<ClientRankingItem[]> {
    return this.request<ClientRankingItem[]>(`/clients/ranking?topCount=${topCount}&days=${days}`)
  }

  // 获取客户端连接热力图数据
  async getClientConnectionHeatmap(days: number = 7): Promise<Record<number, Record<number, Record<string, number>>>> {
    return this.request<Record<number, Record<number, Record<string, number>>>>(`/clients/heatmap?days=${days}`)
  }

  // 获取包含客户端统计的完整仪表板数据（推荐使用）
  async getCombinedDashboardData(days: number = 30): Promise<CombinedDashboardData> {
    return this.request<CombinedDashboardData>(`/dashboard-with-clients?days=${days}`)
  }

  // =================== Channel监控相关API方法 ===================

  // 获取Channel状态
  async getChannelStatus(): Promise<ChannelStatistics> {
    return this.request<ChannelStatistics>('/channel/status')
  }

  // Channel健康检查
  async getChannelHealth(): Promise<ChannelHealthCheck> {
    return this.request<ChannelHealthCheck>('/channel/health')
  }
}

// 创建默认实例
export const statisticsApi = new StatisticsApi()

// React Hook for statistics data
export function useStatistics() {
  const [data, setData] = React.useState<DashboardData | null>(null)
  const [loading, setLoading] = React.useState(false)
  const [error, setError] = React.useState<string | null>(null)

  const fetchDashboardData = React.useCallback(async (days: number = 30) => {
    setLoading(true)
    setError(null)
    
    try {
      const dashboardData = await statisticsApi.getDashboardData(days)
      setData(dashboardData)
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : '获取统计数据失败'
      setError(errorMessage)
      console.error('Failed to fetch dashboard data:', err)
    } finally {
      setLoading(false)
    }
  }, [])

  return {
    data,
    loading,
    error,
    fetchDashboardData,
    refetch: () => fetchDashboardData(30)
  }
}

// React Hook for combined statistics data (包括客户端统计)
export function useCombinedStatistics() {
  const [data, setData] = React.useState<CombinedDashboardData | null>(null)
  const [loading, setLoading] = React.useState(false)
  const [error, setError] = React.useState<string | null>(null)

  const fetchCombinedDashboardData = React.useCallback(async (days: number = 30) => {
    setLoading(true)
    setError(null)
    
    try {
      const dashboardData = await statisticsApi.getCombinedDashboardData(days)
      setData(dashboardData)
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : '获取统计数据失败'
      setError(errorMessage)
      console.error('Failed to fetch combined dashboard data:', err)
    } finally {
      setLoading(false)
    }
  }, [])

  return {
    data,
    loading,
    error,
    fetchCombinedDashboardData,
    refetch: () => fetchCombinedDashboardData(30)
  }
}

// 为兼容性导入React
import * as React from "react"